Set sys.argv[0] in build scripts run by build_meta

aimileus opened the pull request:
<!-- First time contributors: Take a moment to review https://setuptools.readthedocs.io/en/latest/developer-guide.html! -->
<!-- Remove sections if not applicable -->

## Summary of changes

Some setup.py scripts, use sys.argv[0] to locate the source directory of a project. I only added this to build_meta.\_\_legacy__ since that is focused on backwards compatibility with old scripts. However, @pganssle said this behaviour should not be added to setuptools.build_meta. 

Closes #1628 

### Pull Request Checklist
- [x] Changes have tests
- [x] News fragment added in changelog.d. See [documentation](http://setuptools.readthedocs.io/en/latest/developer-guide.html#making-a-pull-request) for details

pganssle commented:
I don't think we actually want to pass through `sys.argv`, we need to mock the `sys.argv` that we would hav executed in the first place.

aimileus commented:
I am not sure I understand? This only sets the *first* item of sys.argv, and keeps the other arguments, which are defined elsewhere (for example in _BuildMetaBackend.build_wheel). sys.argv should be (exactly) the same in the `python3 setup.py bdist_wheel` and `build_meta.__legacy__.build_wheel()` after this PR.

pganssle commented:
@aimileus Oh sorry, I'm a bit distracted and misunderstood this PR. Can you add a test for this?

aimileus commented:
Of course! I added some tests. The other legacy behavior checked that did only work on build_meta.__legacy__ and not on build_meta. I added a similar check. Is that correct?

webknjaz commented:
@aimileus would you mind resolving the conflicts?

aimileus commented:
Sure! I rebased on master.

jaraco commented:
Given this has been reviewed by Paul and Sviatoslav and all requests have been addressed, I'm inclined to accept it. Since this was submitted prior to the Setuptools 45, I'll merge it into maint/44.x.
